### DISCUSSION TITLE
Adding dimensions in paperspace and locate at the right line in a viewport

### DISCUSSION THREAD

**Discussion by Jopie01** [mozman/ezdxf#1181] [Q&A] [answered]
First, thanks for this great library! It's big and I think I overlook something which is already there.

My usecase is as follows:
I'm using CadQuery to create 3D models and put them together into assemblies. After that I want to create some production drawings with some dimensions. As the current DXF export in CadQuery is not very capable, I'm making a lot of changes to it to fit my usecase. I'm currently able to export several drawing views to one DXF in modelspace.

Because we want multiple pages, we need to use paperspace. I've already done a lot of research in that area and I now start with a predefined DXF-file with layers, viewports etc. add my drawing views in modelspace which are then automatically displayed in paperspace.

The problem now is to add the dimensions. I know the points in modelspace relative to the WCS. But in paperspace that WCS is not there. Also the position of the viewport, scale etc have to be taken into account. And that's where I'm lost currently, how to translate my know points in modelspace to the correct position in paperspace?

Does anybody had the same problem or do you have any idea where I can look?

**Comment by mozman** [comment, marked as answer]
The [Viewport.get_transformation_matrix()](https://ezdxf.mozman.at/docs/dxfentities/viewport.html#ezdxf.entities.Viewport.get_transformation_matrix) method returns the transformation matrix from modelspace to paperspace.

Transform a single coordinate by the [Matrix44.transform()](https://ezdxf.mozman.at/docs/math/core.html#ezdxf.math.Matrix44.transform) method or [Matrix44.transform_vertices()](https://ezdxf.mozman.at/docs/math/core.html#ezdxf.math.Matrix44.transform_vertices) to transform multiple coordinates at once.

**Reply by Jopie01** [reply]
Thanks! I knew I was missing something ... I'm using https://github.com/mozman/ezdxf/blob/master/examples/paperspace_setup_R2000.py for testing and I'm able to place a circle around a marker, change the center of the viewport and it still works!

Small problem though .... adding a dimension and transforms it gives an `AttributeError`

```
AttributeError: 'DimStyleOverride' object has no attribute 'transform'
```

code I changed from the file above:
```python
def setup_layout2(doc: Drawing):
    layout2 = doc.layouts.new("scale 1-1")
    # The default paperspace scale is 1:1
    # 1 mm printed is 1 drawing unit in paperspace
    # For most use cases this is the preferred scaling and important fact:
    # the paperspace scaling has no influence on the VIEWPORT scaling - this is
    # a total different topic, see example "viewports_in_paperspace.py"

    layout2.page_setup(size=(297, 210), margins=(10, 10, 10, 10), units="mm")
    viewport = layout2.add_viewport(
        # center of viewport in paperspace units
        center=(100, 100),
        # viewport size in paperspace units
        size=(200, 180),
        # modelspace point to show in center of viewport in WCS
        view_center_point=(25, 30),
        # how much modelspace area to show in viewport in drawing units
        view_height=80,
        status=2,
    )

    matrix = viewport.get_transformation_matrix()

    dim = layout2.add_linear_dim(
        base=(15, 10),  # location of the dimension line
        p1=(10, 10),  # 1st measurement point
        p2=(10, 20),  # 2nd measurement point
        angle=90,
        dimstyle="EZDXF",  # default dimension style
    )
    dim.transform(matrix)   # does not work
    dim.dimension.transform(matrix)  # this works
    dim.render()
```
Looking into the documentation, and indeed, no transform for `DimStyleOverride` but the `Dimension` does.

edit: Luckily the `DimStyleOverride` have a link with `Dimension` through `dim.dimension`. But now the dimension text has a wrong value because of the viewport scaling I suppose. Need to dig into that a bit further, but my biggest problem of placing the dimension correctly is solved!

**Reply by Jopie01** [reply]
Ok, indeed, getting the viewport scale with `viewport.get_scale()` and then apply it to the dimension with the override will show the right value.

```python
scale = viewport.get_scale()

dim = layout2.add_linear_dim(
    base=(15, 10),  # location of the dimension line
    p1=(10, 10),  # 1st measurement point
    p2=(10, 20),  # 2nd measurement point
    angle=90,
    dimstyle="EZDXF",  # default dimension style
    override={
        "dimlfac": 1 / scale
    }
)
dim.dimension.transform(matrix)  # this works
dim.render()
```
